### ISSUE TITLE
I noticed that crossterm::event::read handling of surrogate pairs was broken on Windows,

### ISSUE THREAD

**Issue by tasogare3710** [crossterm-rs/crossterm#787]
I would like to send pull request for this issue but the following issue is blocking it.

I noticed that crossterm::event::read handling of surrogate pairs was broken on Windows, so I fixed it and now this solution lets me know about illegal surrogates. We need to use PPYP pattern since such illegal input can poop(something terrible will happen with illegal input), but i can't decide when to pre-pooping pants. It is not friendly to crash as soon as it receives user input.

Do you have any good ideas?

In my opinion, crossterm::event::read may need a way to notify caller when it receives invalid input.

**Comment by tasogare3710** [comment]
Sorry, it takes time to upload the repository.

**Comment by tasogare3710** [comment]
Sorry, I'm late.

# TL;DR - main issue

Another issue is arises by the fact that the PPYP pattern have to used but cannot fail fast.

- - -

You will know that you received illegal input in the following [places in this file where](https://github.com/tasogare3710/crossterm/blob/fix_handling_of_surrogate_pairs/src/event/source/windows.rs) you call `missing_high_surrogate` and `two_high_surrogates`.

- https://github.com/tasogare3710/crossterm/blob/272dbd76cbd8f9f702d44673c34103ef8e037079/src/event/source/windows.rs#L78
- https://github.com/tasogare3710/crossterm/blob/fix_handling_of_surrogate_pairs/src/event/source/windows.rs#L97
- https://github.com/tasogare3710/crossterm/blob/fix_handling_of_surrogate_pairs/src/event/source/windows.rs#L116
- https://github.com/tasogare3710/crossterm/blob/fix_handling_of_surrogate_pairs/src/event/source/windows.rs#L139
- https://github.com/tasogare3710/crossterm/blob/fix_handling_of_surrogate_pairs/src/event/source/windows.rs#L160
- https://github.com/tasogare3710/crossterm/blob/fix_handling_of_surrogate_pairs/src/event/source/windows.rs#L181


It's hard to restore internal state gracefully when receiving illegal input, so it's ideal to fail fast, but calling panic as soon as the library receives input gives your application lose a chance to clean up. This may make the library less usable. But have to prevent worse things from happening.

In this case, there are two solutions I came up with.

1. function calls panic as soon as it receives illegal input
2. a function returns an error value representing that it received illegal input and prompts crate users to take appropriate action.

1&#41; has the issues already described and 2) has multiple issues with the current crossterm:

# Issue with the solution of 2)

1. Since illegal input can be received on all platforms, a new platform-unspecific error value is needed to represent illegal input.
	- This error has different security semantics from std::io::Error, so it is a prerequisite that the related API can return error types other than std::io::Error.
2. Related API returns std::io::Result
3. Errors in 1) may not be appropriatly handled by crate users.

1&#41; and 2) require API breaking changes. 3) defeats the purpose of PPYP if crate users fails to appropriatly handle errors.

Do you have any ideas for such a case?
